<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="rviz" default="true"/>

  <!-- Parameters -->
  <arg name="config_file" default="omiya_calibration_room_2023.param.yaml"/>
  <arg name="publish_tfs" default="true"/>
  <arg name="fixed_ground_plane_model" default="false"/>

  <!-- Frame configuration -->
  <let name="base_frame" value="base_link"/>
  <let name="main_calibration_sensor_frame" value="pandar_40p_left"/>

  <let name="calibration_lidar_0_frame" value="pandar_40p_left"/>
  <let name="calibration_lidar_1_frame" value="pandar_40p_right"/>
  <let name="calibration_lidar_2_frame" value="pandar_40p_front"/>
  <let name="calibration_lidar_3_frame" value="pandar_40p_rear"/>

  <let name="calibration_camera_0_frame" value="camera0/camera_optical_link"/>
  <let name="calibration_camera_1_frame" value="camera1/camera_optical_link"/>
  <let name="calibration_camera_2_frame" value="camera2/camera_optical_link"/>
  <let name="calibration_camera_3_frame" value="camera3/camera_optical_link"/>
  <let name="calibration_camera_4_frame" value="camera4/camera_optical_link"/>
  <let name="calibration_camera_5_frame" value="camera5/camera_optical_link"/>
  <let name="calibration_camera_6_frame" value="camera6/camera_optical_link"/>

  <!-- Lidar configuration -->
  <let name="calibration_lidar_0_model" value="pandar_40p"/>
  <let name="calibration_lidar_1_model" value="pandar_40p"/>
  <let name="calibration_lidar_2_model" value="pandar_40p"/>
  <let name="calibration_lidar_3_model" value="pandar_40p"/>

  <!-- Topic configuration -->
  <let name="calibration_lidar_0_topic" value="/sensing/lidar/left_upper/pointcloud_raw"/>
  <let name="calibration_lidar_1_topic" value="/sensing/lidar/right_upper/pointcloud_raw"/>
  <let name="calibration_lidar_2_topic" value="/sensing/lidar/front_lower/pointcloud_raw"/>
  <let name="calibration_lidar_3_topic" value="/sensing/lidar/rear_lower/pointcloud_raw"/>

  <let name="calibration_camera_info_0_topic" value="/sensing/camera/camera0/camera_info"/>
  <let name="calibration_camera_info_1_topic" value="/sensing/camera/camera1/camera_info"/>
  <let name="calibration_camera_info_2_topic" value="/sensing/camera/camera2/camera_info"/>
  <let name="calibration_camera_info_3_topic" value="/sensing/camera/camera3/camera_info"/>
  <let name="calibration_camera_info_4_topic" value="/sensing/camera/camera4/camera_info"/>
  <let name="calibration_camera_info_5_topic" value="/sensing/camera/camera5/camera_info"/>
  <let name="calibration_camera_info_6_topic" value="/sensing/camera/camera6/camera_info"/>

  <let name="calibration_compressed_image_0_topic" value="/sensing/camera/camera0/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_1_topic" value="/sensing/camera/camera1/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_2_topic" value="/sensing/camera/camera2/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_3_topic" value="/sensing/camera/camera3/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_4_topic" value="/sensing/camera/camera4/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_5_topic" value="/sensing/camera/camera5/image_rect_color/compressed"/>
  <let name="calibration_compressed_image_6_topic" value="/sensing/camera/camera6/image_rect_color/compressed"/>

  <arg
    name="auxiliar_tag_ids"
    default="[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]"
  />
  <arg name="waypoint_tag_ids" default="[0, 1, 2, 3, 4, 5]"/>
  <arg
    name="ground_tag_ids"
    default="[50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]"
  />
  <arg name="left_wheel_tag_id" default="22"/>
  <arg name="right_wheel_tag_id" default="26"/>

  <!-- tag_based_sfm_base_lidar calibrator -->
  <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/calibrator.launch.xml">
    <arg name="calibration_service_name" value="calibrate_base_lidars_cameras"/>
    <arg name="publish_tfs" value="$(var publish_tfs)"/>
    <arg name="fixed_ground_plane_model" value="$(var fixed_ground_plane_model)"/>

    <arg name="base_frame" value="$(var base_frame)"/>
    <arg name="main_calibration_sensor_frame" value="$(var main_calibration_sensor_frame)"/>

    <arg name="calibration_lidar_0_frame" value="$(var calibration_lidar_0_frame)"/>
    <arg name="calibration_lidar_1_frame" value="$(var calibration_lidar_1_frame)"/>
    <arg name="calibration_lidar_2_frame" value="$(var calibration_lidar_2_frame)"/>
    <arg name="calibration_lidar_3_frame" value="$(var calibration_lidar_3_frame)"/>

    <arg name="calibration_camera_0_frame" value="$(var calibration_camera_0_frame)"/>
    <arg name="calibration_camera_1_frame" value="$(var calibration_camera_1_frame)"/>
    <arg name="calibration_camera_2_frame" value="$(var calibration_camera_2_frame)"/>
    <arg name="calibration_camera_3_frame" value="$(var calibration_camera_3_frame)"/>
    <arg name="calibration_camera_4_frame" value="$(var calibration_camera_4_frame)"/>
    <arg name="calibration_camera_5_frame" value="$(var calibration_camera_5_frame)"/>
    <arg name="calibration_camera_6_frame" value="$(var calibration_camera_6_frame)"/>

    <arg name="calibration_lidar_0_model" value="$(var calibration_lidar_0_model)"/>
    <arg name="calibration_lidar_1_model" value="$(var calibration_lidar_1_model)"/>
    <arg name="calibration_lidar_2_model" value="$(var calibration_lidar_2_model)"/>
    <arg name="calibration_lidar_3_model" value="$(var calibration_lidar_3_model)"/>

    <arg name="calibration_lidar_0_topic" value="$(var calibration_lidar_0_topic)"/>
    <arg name="calibration_lidar_1_topic" value="$(var calibration_lidar_1_topic)"/>
    <arg name="calibration_lidar_2_topic" value="$(var calibration_lidar_2_topic)"/>
    <arg name="calibration_lidar_3_topic" value="$(var calibration_lidar_3_topic)"/>

    <arg name="calibration_camera_info_0_topic" value="$(var calibration_camera_info_0_topic)"/>
    <arg name="calibration_camera_info_1_topic" value="$(var calibration_camera_info_1_topic)"/>
    <arg name="calibration_camera_info_2_topic" value="$(var calibration_camera_info_2_topic)"/>
    <arg name="calibration_camera_info_3_topic" value="$(var calibration_camera_info_3_topic)"/>
    <arg name="calibration_camera_info_4_topic" value="$(var calibration_camera_info_4_topic)"/>
    <arg name="calibration_camera_info_5_topic" value="$(var calibration_camera_info_5_topic)"/>
    <arg name="calibration_camera_info_6_topic" value="$(var calibration_camera_info_6_topic)"/>

    <arg name="calibration_compressed_image_0_topic" value="$(var calibration_compressed_image_0_topic)"/>
    <arg name="calibration_compressed_image_1_topic" value="$(var calibration_compressed_image_1_topic)"/>
    <arg name="calibration_compressed_image_2_topic" value="$(var calibration_compressed_image_2_topic)"/>
    <arg name="calibration_compressed_image_3_topic" value="$(var calibration_compressed_image_3_topic)"/>
    <arg name="calibration_compressed_image_4_topic" value="$(var calibration_compressed_image_4_topic)"/>
    <arg name="calibration_compressed_image_5_topic" value="$(var calibration_compressed_image_5_topic)"/>
    <arg name="calibration_compressed_image_6_topic" value="$(var calibration_compressed_image_6_topic)"/>

    <arg name="auxiliar_tag_ids" value="$(var auxiliar_tag_ids)"/>
    <arg name="waypoint_tag_ids" value="$(var waypoint_tag_ids)"/>
    <arg name="ground_tag_ids" value="$(var ground_tag_ids)"/>
    <arg name="left_wheel_tag_id" value="$(var left_wheel_tag_id)"/>
    <arg name="right_wheel_tag_id" value="$(var right_wheel_tag_id)"/>
  </include>
</launch>
